packages/machine-extractor/src/actions.ts

```diff
@@ -99,6 +99,52 @@ export const ActionAsString = maybeTsAsExpression(
   ),
 );
 
+// List of builtin XState actions that are treated specially in Studio UI
+const SUPPORTED_BUILTIN_ACTIONS = [
+  'xstate.assign',
+  'xstate.log',
+  'xstate.raise',
+  'xstate.stop',
+  'xstate.sendTo',
+];
+
+/**
+ * {type: 'custom name', params: {foo: 'bar'}} Must be extracted as named action
+ * {type: someIdentifier, params: {foo: 'bar'}} Must be extracted as inline action
+ * {type: 'xstate.assign', assignment: {foo: 'bar', baz: () => {}}} Must be extracted as inline action
+ */
+export const ActionAsObjectExpression = createParser({
+  babelMatcher: t.isObjectExpression,
+  parseNode: (node, context): ActionNode => {
+    const id = context.getNodeHash(node);
+    for (const prop of node.properties) {
+      if (t.isObjectProperty(prop)) {
+        if (
+          t.isIdentifier(prop.key) &&
+          prop.key.name === 'type' &&
+          t.isStringLiteral(prop.value) &&
+          !SUPPORTED_BUILTIN_ACTIONS.includes(prop.value.value)
+        ) {
+          return {
+            action: id,
+            node,
+            name: prop.value.value,
+            declarationType: 'object',
```

Comment:
imho this could just be `declarationType: 'named'` and if `node` turns out to be an object expression then its properties should get extracted~ in `getActionConfig`

Reply:
The check to determine whether it's an object that needs to be specially treated is more complicated than that. It checks if it's an object that has a `type` property that has a string value and the value is not a built-in XState action name that we support in the Studio UI.